### PR TITLE
[POAE7-2913] Vectorized Filter Support

### PR DESCRIPTION
**Pull request by bigPYJ1151**
### What changes were proposed in this pull request?

Support vectorized filter.

### Why are the changes needed?

Performance optimization.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UTs

### Which label does this PR belong to?

INFRA

### CODE REVIEW

File: cpp/src/cider/exec/nextgen/context/CodegenContext.cpp [hunk 1]
```diff
@@ -386,5 +393,17 @@ void bitBufferAnd(jitlib::JITValuePointer& output,
           .params_vector = {output.get(), a.get(), b.get(), bit_num.get()}});
 }
 
+void convertByteBoolToBit(jitlib::JITValuePointer& byte,
+                          jitlib::JITValuePointer& bit,
+                          jitlib::JITValuePointer& len) {
+  CHECK((isaTypedJITValue<JITTypeTag::POINTER, JITTypeTag::INT8>(byte)));
```
Comment (harborn) [inline cpp/src/cider/exec/nextgen/context/CodegenContext.cpp:399]: if CHECK failed, FATAL message log will print, and application will abort.
Is that your are expected?

Comment (bigPYJ1151) [reply cpp/src/cider/exec/nextgen/context/CodegenContext.cpp:399]: Yes, I expect all potential errors of the code generation will expose explicitly in tests.

File: cpp/src/cider/exec/nextgen/transformer/Transformer.cpp [hunk 2]
```diff
@@ -115,10 +117,13 @@ TranslatorPtr Transformer::toTranslator(OpPipeline& pipeline,
...
-  // Vectorize Project Transformation
+  // Vectorize Project and Filter Transformation
+  // Currently, auto-vectorize will be applied to pure project pipelines or filter
+  // pipelines.
   auto traverse_pivot = ++pipeline.begin();
-  if (co.enable_vectorize && isa<ProjectNode>(*traverse_pivot)) {
-    // Currently, auto-vectorize will be applied to pure project pipeline only.
+  if (co.enable_vectorize &&
+      (isa<ProjectNode>(*traverse_pivot) || isa<FilterNode>(*traverse_pivot))) {
+    bool has_filter = isa<FilterNode>(*traverse_pivot);
```
[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply cpp/src/cider/exec/nextgen/transformer/Transformer.cpp:126]: Exactly.

File: cpp/src/cider/exec/nextgen/operators/VectorizedFilterNode.cpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+using namespace jitlib;
+
+TranslatorPtr VectorizedFilterNode::toTranslator(const TranslatorPtr& successor) {
+  return createOpTranslator<VectorizedFilterTranslator>(shared_from_this(), successor);
+}
+
+jitlib::JITValuePointer VectorizedFilterTranslator::generateFilterCondition(
+    CodegenContext& context) {
+  auto&& [_, exprs] = node_->getOutputExprs();
+  // Combine conditions with AND.
```
[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply cpp/src/cider/exec/nextgen/operators/VectorizedFilterNode.cpp:48]: The rule is hard to define.

File: cpp/src/cider/exec/nextgen/operators/RowToColumnNode.cpp [hunk 4]
```diff
@@ -458,9 +458,7 @@ void RowToColumnTranslator::codegenImpl(SuccessorEmitter successor_wrapper,
 
   auto output_index = func->createVariable(JITTypeTag::INT64, "output_index", 0);
 
-  // Get input ArrowArray length from previous C2RNode
-  auto prev_c2r_node = static_cast<RowToColumnNode*>(node_.get())->getColumnToRowNode();
-  auto input_array_len = prev_c2r_node->getColumnRowNum();
+  auto input_array_len = context.getInputLength();
```
[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply cpp/src/cider/exec/nextgen/operators/RowToColumnNode.cpp:461]: The length of input will never be changed.

File: cpp/src/cider/exec/nextgen/operators/VectorizedFilterNode.cpp [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+        // Set current i64 selected mask.
+        context.setFilterMask(current_selected_mask, row_index_start);
+
+        // All the rows are filtered.
+        builder->loopContinue(current_selected_mask == 0);
+
+        // TODO (bigPYJ1151): Support full execution when no rows are filtered.
+        // Row-based execution.
+        successor_wrapper(successor, context);
```
[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply cpp/src/cider/exec/nextgen/operators/VectorizedFilterNode.cpp:130]: Yes.

File: cpp/src/cider/exec/nextgen/operators/VectorizedFilterNode.cpp [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  // Register batch for the condition.
+  auto condition_array = context.registerBatch(conditon->get_type_info(), "condition");
+  conditon->setLocalIndex(context.appendArrowArrayValues(
+      condition_array, utils::JITExprValue(0, JITExprValueType::BATCH)));
+
+  // Generate the condition.
+  ExprPtrVector condition_target{conditon};
+  auto vectorized_proj_node = createOpNode<VectorizedProjectNode>(condition_target);
```
[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply cpp/src/cider/exec/nextgen/operators/VectorizedFilterNode.cpp:69]: Sounds good.

File: cpp/src/cider/exec/nextgen/operators/VectorizedProjectNode.cpp [hunk 7]
```diff
@@ -480,10 +481,11 @@ void generateBoolInputExprsGroupCode(context::CodegenContext& context,
   auto input_len = context.getInputLength();
 
   auto bool_loop_length = (input_len + 7) / 8;
+  auto buffer_bytes_length = (bool_loop_length + 31) / 32 * 32;
```
Comment (jikunshang) [inline cpp/src/cider/exec/nextgen/operators/VectorizedProjectNode.cpp:484]: / 32 * 32 => >>5  << 5

File: cpp/src/cider/tests/CiderBitUtilsTest.cpp [hunk 8]
```diff
@@ -170,6 +170,14 @@ TEST_F(CiderBitUtilsTest, ByteToBitTest) {
   check_byte_bit(byte_vec_backup.data(), bit_vec.data(), byte_vec_backup.size());
 }
 
+TEST_F(CiderBitUtilsTest, CTZTest) {
+  size_t data = 0xFFFFFFFFFFFFFFFF;
+  for (size_t i = 0; i < 63; ++i) {
```
Comment (jikunshang) [inline cpp/src/cider/tests/CiderBitUtilsTest.cpp:175]: i< 64?

**Review by jikunshang — APPROVED** [review]

**Review by Ma-Jian1 — APPROVED** [review]

**Pull request merged**